home-manager-module.nix

```diff
@@ -0,0 +1,100 @@
+{ systemdLib }:
+{
+  config,
+  osConfig,
+  options,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.virtualisation.user.quadlet;
+  quadletUtils = import ./utils.nix {
+    inherit lib;
+    systemdLib = systemdLib { inherit lib config pkgs; };
+  };
+  containerOpts = lib.types.submodule (import ./container.nix { inherit quadletUtils; });
+  networkOpts = lib.types.submodule (import ./network.nix { inherit quadletUtils pkgs; });
+in
+{
+  options.virtualisation.user.quadlet = {
+    autoUpdate = {
+      enable = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+      };
+      calendar = lib.mkOption {
+        type = lib.types.str;
+        default = "*-*-* 00:00:00";
+      };
+    };
+    containers = lib.mkOption {
+      type = lib.types.attrsOf containerOpts;
+      default = { };
+    };
+    networks = lib.mkOption {
+      type = lib.types.attrsOf networkOpts;
+      default = { };
+    };
+  };
+  config =
+    let
+      allObjects = (lib.attrValues cfg.containers) ++ (lib.attrValues cfg.networks);
+    in
+    {
+      xdg.configFile =
+        let
+          links = pkgs.linkFarm "user-quadlet-service-symlinks" (
+            map (p: {
+              name = p._unitName;
+              path = "/run/user/${
+                toString osConfig.users.users.${config.home.username}.uid
+              }/systemd/generator/${p._unitName}";
+            }) allObjects
+          );
+        in
+        lib.mergeAttrsList (
+          map (p: {
+            # Install the .container, .network, etc files
+            "containers/systemd/${p._configName}" = {
+              text = p._configText;
+            };
+            # Link the corresponding .service files so that the home-manager activation process knows about them
+            "systemd/user/${p._unitName}" = {
+              source = "${links}/${p._unitName}";
```

Comment:
Is it possible to link directly from /run/user/... without linkFarm? With that we might even be able to use `$UID` without relying on `uid` being explicitly set in nixos config.

Reply:
Yea, this bit I'm not 100% happy with either. The symlinks are definitely needed so that the home manager reactivation can restart them. But while home-manager let's us write files into $HOME, it doesn't let us write symlinks directly (as far as I can find). So I used `linkFarm` which creates a derivation with links from `generated-service-name.service` to `/run/user/UID/systemd/generator/generated-service-name.service`

Then home-manager creates a link to the `linkFarm` file (which is itself a link). The end result is a chain of links:

```
~/.config/systemd/user/servicename.service
-> /nix/store/sn55a0wl1mxy0fnq74mqrpv0jbc97309-home-manager-files/.config/systemd/user/servicename.service
-> /nix/store/mm66x07g2jld7497s7pf8cfl3l948p4c-user-quadlet-service-symlinks/servicename.service
-> /run/user/3008/systemd/generator/servicename.service
```

Happy to improve it, I'm just not sure how at this point.

Reply:
On second thought perhaps using home-manager's `systemd.user.tmpfiles.rules` would work? We'd still have to reference the user's UID somehow.

Reply:
I gave systemd tmpfiles a try (`"L+ ${config.xdg.configHome}/foo - - - - %t/foo"`), it worked well creating the links but I can't make it delete those link when the config is removed.

I think linkFarm is fine here but getting uid from config isn't ideal as not everyone sets it. Since I don't have a good solution either (except for resorting to `pkgs.runCommandLocal` or activation script), I'm ok to merge the current version for now provided we make it a documented limitation (comment and README).